File: tests/integration-tests/src/test/java/com/jaynewstrom/json/sample/optional/OptionalTest.java
package com.jaynewstrom.json.sample.optional;

import com.jaynewstrom.json.sample.JsonTestHelper;

import org.junit.Test;

import static org.fest.assertions.api.Assertions.assertThat;

public final class OptionalTest {
    @Test public void whenOneIsNull_EnsureTwoIsParsed() {
        Optional optional = new JsonTestHelper().deserializeFile(Optional.class, "OneNull.json", this);
        assertThat(optional.one).isNull();
        assertThat(optional.two).isEqualTo("foo");
    }

    @Test public void whenTwoIsNull_EnsureOneIsParsed() {
        Optional optional = new JsonTestHelper().deserializeFile(Optional.class, "TwoNull.json", this);
        assertThat(optional.one).isEqualTo("foo");
        assertThat(optional.two).isNull();
    }

    @Test public void ensureBothCanBeNull() {
        Optional optional = new JsonTestHelper().deserializeFile(Optional.class, "BothNull.json", this);
        assertThat(optional.one).isNull();
        assertThat(optional.two).isNull();
    }

    @Test public void ensureNullIsSerialized() {
        String result = new JsonTestHelper().serialize(new Optional(null, "Two"));
        assertThat(result).isEqualTo("{\"one\":null,\"two\":\"Two\"}");
    }
}

File: tests/integration-tests/src/test/java/com/jaynewstrom/json/sample/optional/OptionalTest.kt
package com.jaynewstrom.json.sample.optional

import com.jaynewstrom.json.sample.JsonTestHelper
import org.fest.assertions.api.Assertions.assertThat
import org.junit.Test

class OptionalTest {
    @Test fun whenOneIsNull_EnsureTwoIsParsed() {
        val (one, two) = JsonTestHelper().deserializeFile(Optional::class.java, "OneNull.json", this)
        assertThat(one).isNull()
        assertThat(two).isEqualTo("foo")
    }

    @Test fun whenTwoIsNull_EnsureOneIsParsed() {
        val (one, two) = JsonTestHelper().deserializeFile(Optional::class.java, "TwoNull.json", this)
        assertThat(one).isEqualTo("foo")
        assertThat(two).isNull()
    }

    @Test fun ensureBothCanBeNull() {
        val (one, two) = JsonTestHelper().deserializeFile(Optional::class.java, "BothNull.json", this)
        assertThat(one).isNull()
        assertThat(two).isNull()
    }

    @Test fun ensureNullIsSerialized() {
        val result = JsonTestHelper().serialize(Optional(null, "Two"), Optional::class.java)
        assertThat(result).isEqualTo("{\"one\":null,\"two\":\"Two\"}")
    }
}
